pkg/operator/configobserver/node/observe_latency_profile.go

```diff
@@ -0,0 +1,138 @@
+	observeLatencyProfileFn := func(
+		genericListers configobserver.Listers,
+		eventRecorder events.Recorder,
+		existingConfig map[string]interface{},
+	) (ret map[string]interface{}, errs []error) {
+
+		// traverse among different worker latency profiles and their required arg val pairs
+		// and obtain a unique list of paths that are to be used with the observedConfig
+		uniqueNestedPaths := getUniqueNestedPaths(latencyProfileValMap)
+
+		defer func() {
+			// Prune the observed config so that it only contains fields specific to this observer
+			ret = configobserver.Pruned(ret, uniqueNestedPaths...)
+
+			// in case of existingConfig and currently observedConfig has mismatch in fields that are
+			// being monitored eg. apiServerArguments.default-not-ready-toleration-seconds,
+			// extendedArguments.node-monitor-grace-period, etc.
+			prevRet := configobserver.Pruned(existingConfig, uniqueNestedPaths...)
+			if !reflect.DeepEqual(prevRet, ret) {
+				eventRecorder.Eventf("LatencyProfileUpdated", "latency profile on cluster was updated, observed-config = %v", ret)
+			}
+		}()
+
+		listers := genericListers.(NodeLister)
+		configNode, err := listers.NodeLister().Get("cluster")
+		// we got an error so without the node object we are not able to determine worker latency profile
+		if err != nil {
+			// if config/v1/node/cluster object is not found, that can be treated as a non-error case
+			if apierrors.IsNotFound(err) {
+				klog.Warningf("nodes.config.openshift.io/cluster not found")
+			} else {
+				errs = append(errs, err)
+			}
+			return existingConfig, errs
+		}
+
+		cond, err := conditionFn()
+		if err != nil {
+			errs = append(errs, err)
+			return existingConfig, errs
+		}
+
+		// condition is not satisfied, so return existing config back without any changes
+		// eg. conditionFn is to check if rollouts are in progressing state
+		// typically when WorkerLatencyProfileProgressing status (i.e. sprofile rollout is already in progressing state)
+		if !cond {
+			// log that latency profile couldn't be updated due to falsy condition
+			klog.Warningf("latency profile config observer did not update observed config due to unsatisified pre-requisite")
+			return existingConfig, errs
+		}
+
+		if configNode.Spec.WorkerLatencyProfile == "" {
+			// log that latency profile is empty
+			klog.Infof("worker latency profile is not set in nodes.config.openshift.io/cluster")
+			return existingConfig, errs
+		}
+
+		// get the arg val pairs for currently set latency profile
+		profileArgValPairs, ok := latencyProfileValMap[configNode.Spec.WorkerLatencyProfile]
+		if !ok {
+			// log that this is an unsupported latency profile
+			klog.Warningf("unsupported worker latency profile found in nodes.config.openshift.io/cluster Spec.WorkerLatencyProfile = %v", configNode.Spec.WorkerLatencyProfile)
```

Comment:
I would document that the list of allowed values is defined on the API server and we can only accept new values during an upgrade. In that case, we tolerate skew until a new version of the operator will be installed.

Reply:
Documented in comments wherever relevant.
```
// Note: In case new latency profiles are added in the future in openshift/api
// this could break cluster upgrades although we're not passing an error here
// to ensure that ConfigObservationController doesn't reach degraded state.
```